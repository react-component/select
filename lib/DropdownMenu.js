'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _reactDom = require('react-dom');

var _util = require('./util');

var _rcMenu = require('rc-menu');

var _rcMenu2 = _interopRequireDefault(_rcMenu);

var _domScrollIntoView = require('dom-scroll-into-view');

var _domScrollIntoView2 = _interopRequireDefault(_domScrollIntoView);

var DropdownMenu = _react2['default'].createClass({
  displayName: 'DropdownMenu',

  propTypes: {
    prefixCls: _react2['default'].PropTypes.string,
    menuItems: _react2['default'].PropTypes.any,
    search: _react2['default'].PropTypes.any
  },

  componentDidMount: function componentDidMount() {
    this.scrollActiveItemToView();
  },

  shouldComponentUpdate: function shouldComponentUpdate(nextProps) {
    // freeze when hide
    return nextProps.visible;
  },

  componentDidUpdate: function componentDidUpdate(prevProps) {
    var props = this.props;
    if (!prevProps.visible && props.visible) {
      this.scrollActiveItemToView();
    }
  },

  scrollActiveItemToView: function scrollActiveItemToView() {
    // scroll into view
    var itemComponent = (0, _reactDom.findDOMNode)(this.firstActiveItem);
    if (itemComponent) {
      (0, _domScrollIntoView2['default'])(itemComponent, (0, _reactDom.findDOMNode)(this.refs.menu), {
        onlyScrollIfNeeded: true
      });
    }
  },

  renderMenu: function renderMenu() {
    var _this = this;

    var props = this.props;
    var menuItems = props.menuItems;
    var defaultActiveFirstOption = props.defaultActiveFirstOption;
    var value = props.value;
    var dropdownMenuStyle = props.dropdownMenuStyle;
    var prefixCls = props.prefixCls;
    var multiple = props.multiple;
    var onMenuDeselect = props.onMenuDeselect;
    var onMenuSelect = props.onMenuSelect;

    if (menuItems && menuItems.length) {
      var _ret = (function () {
        var menuProps = {};
        if (multiple) {
          menuProps.onDeselect = onMenuDeselect;
          menuProps.onSelect = onMenuSelect;
        } else {
          menuProps.onClick = onMenuSelect;
        }
        var selectedKeys = (0, _util.getSelectKeys)(menuItems, value);
        var activeKeyProps = {};

        var clonedMenuItems = menuItems;
        if (selectedKeys.length) {
          (function () {
            activeKeyProps.activeKey = selectedKeys[0];
            var foundFirst = false;
            // set firstActiveItem via cloning menus
            // for scroll into view
            var clone = function clone(item) {
              if (!foundFirst && selectedKeys.indexOf(item.key) !== -1) {
                foundFirst = true;
                return (0, _react.cloneElement)(item, {
                  ref: function ref(_ref) {
                    _this.firstActiveItem = _ref;
                  }
                });
              }
              return item;
            };

            clonedMenuItems = menuItems.map(function (item) {
              if (item.type === _rcMenu.ItemGroup) {
                var children = item.props.children.map(clone);
                return (0, _react.cloneElement)(item, {}, children);
              }
              return clone(item);
            });
          })();
        }

        return {
          v: _react2['default'].createElement(
            _rcMenu2['default'],
            _extends({
              ref: 'menu',
              defaultActiveFirst: defaultActiveFirstOption,
              style: dropdownMenuStyle
            }, activeKeyProps, {
              multiple: multiple,
              focusable: false
            }, menuProps, {
              selectedKeys: selectedKeys,
              prefixCls: prefixCls + '-menu' }),
            clonedMenuItems
          )
        };
      })();

      if (typeof _ret === 'object') return _ret.v;
    }
    return null;
  },

  render: function render() {
    return _react2['default'].createElement(
      'div',
      null,
      this.props.search,
      this.renderMenu()
    );
  }
});

exports['default'] = DropdownMenu;
module.exports = exports['default'];