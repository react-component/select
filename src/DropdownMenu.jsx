import React, { cloneElement } from 'react';
import { findDOMNode } from 'react-dom';
import PropTypes from 'prop-types';
import toArray from 'rc-util/lib/Children/toArray';
import Menu from 'rc-menu';
import scrollIntoView from 'dom-scroll-into-view';
import { getSelectKeys, preventDefaultEvent, saveRef } from './util';

export default class DropdownMenu extends React.Component {
  static propTypes= {
    defaultActiveFirstOption: PropTypes.bool,
    value: PropTypes.any,
    dropdownMenuStyle: PropTypes.object,
    multiple: PropTypes.bool,
    onPopupFocus: PropTypes.func,
    onMenuDeSelect: PropTypes.func,
    onMenuSelect: PropTypes.func,
    prefixCls: PropTypes.string,
    menuItems: PropTypes.any,
    inputValue: PropTypes.string,
    visible: PropTypes.bool,
  }

  constructor(props) {
    super(props);
    this.saveMenuRef = saveRef.bind(this, 'menuInstance');
  }

  componentWillMount() {
    this.lastInputValue = this.props.inputValue;
  }

  componentDidMount() {
    this.scrollActiveItemToView();
    this.lastVisible = this.props.visible;
  }

  shouldComponentUpdate(nextProps) {
    if (!nextProps.visible) {
      this.lastVisible = false;
    }
    // freeze when hide
    return nextProps.visible;
  }

  componentDidUpdate(prevProps) {
    const props = this.props;
    if (!prevProps.visible && props.visible) {
      this.scrollActiveItemToView();
    }
    this.lastVisible = props.visible;
    this.lastInputValue = props.inputValue;
  }

  scrollActiveItemToView = () => {
    // scroll into view
    const itemComponent = findDOMNode(this.firstActiveItem);
    const props = this.props;

    if (itemComponent) {
      const scrollIntoViewOpts = {
        onlyScrollIfNeeded: true,
      };
      if ((!props.value || props.value.length === 0) && props.firstActiveValue) {
        scrollIntoViewOpts.alignWithTop = true;
      }

      scrollIntoView(itemComponent, findDOMNode(this.menuInstance), scrollIntoViewOpts);
    }
  }

  renderMenu() {
    const props = this.props;
    const {
      menuItems,
      defaultActiveFirstOption, value,
      prefixCls, multiple,
      onMenuSelect, inputValue, firstActiveValue,
    } = props;
    if (menuItems && menuItems.length) {
      const menuProps = {};
      if (multiple) {
        menuProps.onDeselect = props.onMenuDeselect;
        menuProps.onSelect = onMenuSelect;
      } else {
        menuProps.onClick = onMenuSelect;
      }

      const selectedKeys = getSelectKeys(menuItems, value);
      const activeKeyProps = {};

      let clonedMenuItems = menuItems;
      if (selectedKeys.length || firstActiveValue) {
        if (props.visible && !this.lastVisible) {
          activeKeyProps.activeKey = selectedKeys[0] || firstActiveValue;
        }
        let foundFirst = false;
        // set firstActiveItem via cloning menus
        // for scroll into view
        const clone = (item) => {
          if ((!foundFirst && selectedKeys.indexOf(item.key) !== -1)
            || (!foundFirst && !selectedKeys.length && firstActiveValue.indexOf(item.key) !== -1)) {
            foundFirst = true;
            return cloneElement(item, {
              ref: (ref) => {
                this.firstActiveItem = ref;
              },
            });
          }
          return item;
        };

        clonedMenuItems = menuItems.map(item => {
          if (item.type.isMenuItemGroup) {
            const children = toArray(item.props.children).map(clone);
            return cloneElement(item, {}, children);
          }
          return clone(item);
        });
      }


      // clear activeKey when inputValue change
      if (inputValue !== this.lastInputValue) {
        activeKeyProps.activeKey = '';
      }

      return (<Menu
        ref={this.saveMenuRef}
        style={this.props.dropdownMenuStyle}
        defaultActiveFirst={defaultActiveFirstOption}
        {...activeKeyProps}
        multiple={multiple}
        focusable={false}
        {...menuProps}
        selectedKeys={selectedKeys}
        prefixCls={`${prefixCls}-menu`}
      >
        {clonedMenuItems}
      </Menu>);
    }
    return null;
  }

  render() {
    const renderMenu = this.renderMenu();
    return renderMenu ? (
      <div
        style={{ overflow: 'auto' }}
        onFocus={this.props.onPopupFocus}
        onMouseDown={preventDefaultEvent}
      >
        {renderMenu}
      </div>
    ) : null;
  }
}

DropdownMenu.displayName = 'DropdownMenu';
