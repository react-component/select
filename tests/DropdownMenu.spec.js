/* eslint-disable no-undef */
import React from 'react';
import { mount, render } from 'enzyme';
import { Item as MenuItem, ItemGroup as MenuItemGroup } from 'rc-menu';
import DropdownMenu from '../src/DropdownMenu';

describe('DropdownMenu', () => {
  it('renders correctly', () => {
    const menuItems = [
      <MenuItemGroup key="group1">
        <MenuItem key="1">1</MenuItem>
      </MenuItemGroup>,
      <MenuItemGroup key="group2">
        <MenuItem key="2">2</MenuItem>
      </MenuItemGroup>,
    ];

    const wrapper = render(<DropdownMenu menuItems={menuItems} value={['1']} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('save first active item', () => {
    const menuItems = [<MenuItem key="1">1</MenuItem>, <MenuItem key="2">2</MenuItem>];

    const wrapper = mount(<DropdownMenu menuItems={menuItems} value={['1']} />);

    expect(wrapper.instance().firstActiveItem.props.children).toBe('1');
  });

  it('set active key to menu', () => {
    const menuItems = [<MenuItem key="1">1</MenuItem>, <MenuItem key="2">2</MenuItem>];

    const wrapper = mount(<DropdownMenu menuItems={menuItems} value={['1']} />);

    wrapper.setProps({ visible: true });
    expect(wrapper.find('Menu').props().activeKey).toBe('1');

    wrapper.setProps({ inputValue: 'foo' });
    expect(wrapper.find('Menu').props().activeKey).toBe('');
  });

  it('save firstActiveValue', () => {
    const menuItems = [<MenuItem key="1">1</MenuItem>, <MenuItem key="2">2</MenuItem>];

    const wrapper = mount(<DropdownMenu menuItems={menuItems} firstActiveValue="2" />);

    expect(wrapper.instance().firstActiveItem.props.children).toBe('2');
  });

  it('set firstActiveValue key to menu', () => {
    const menuItems = [<MenuItem key="1">1</MenuItem>, <MenuItem key="2">2</MenuItem>];

    const wrapper = mount(<DropdownMenu menuItems={menuItems} firstActiveValue="2" />);

    wrapper.setProps({ visible: true });
    expect(wrapper.find('Menu').props().activeKey).toBe('2');

    wrapper.setProps({ inputValue: 'foo' });
    expect(wrapper.find('Menu').props().activeKey).toBe('');
  });

  it('save value not firstActiveValue', () => {
    const menuItems = [<MenuItem key="1">1</MenuItem>, <MenuItem key="2">2</MenuItem>];

    const wrapper = mount(
      <DropdownMenu value={['1']} menuItems={menuItems} firstActiveValue="2" />,
    );

    expect(wrapper.instance().firstActiveItem.props.children).toBe('1');
  });

  it('save visible and inputValue when update', () => {
    const wrapper = mount(<DropdownMenu />);

    wrapper.setProps({ visible: true, inputValue: 'foo' });

    expect(wrapper.instance().lastVisible).toBe(true);
    expect(wrapper.instance().lastInputValue).toBe('foo');
  });

  it('not update when next visible is false', () => {
    const wrapper = mount(<DropdownMenu />);

    expect(wrapper.instance().shouldComponentUpdate({ visible: true })).toBe(true);
    expect(wrapper.instance().shouldComponentUpdate({ visible: false })).toBe(false);
  });

  it('should updated when input value change', () => {
    const wrapper = mount(<DropdownMenu visible={false} inputValue="test" />);

    expect(wrapper.instance().shouldComponentUpdate({ inputValue: 'test2' })).toBe(true);
  });
});
