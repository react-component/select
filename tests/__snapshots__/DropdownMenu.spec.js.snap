// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropdownMenu renders correctly 1`] = `
<div
  style="overflow:auto"
>
  <ul
    class="undefined-menu  undefined-menu-root undefined-menu-vertical"
    role="listbox"
    tabindex="0"
  >
    <li
      class=" undefined-menu-item-group"
    >
      <div
        class="undefined-menu-item-group-title"
      />
      <ul
        class="undefined-menu-item-group-list"
      >
        <li
          class="undefined-menu-item undefined-menu-item-selected"
          role="menuitem"
        >
          1
        </li>
      </ul>
    </li>
    <li
      class=" undefined-menu-item-group"
    >
      <div
        class="undefined-menu-item-group-title"
      />
      <ul
        class="undefined-menu-item-group-list"
      >
        <li
          class="undefined-menu-item"
          role="menuitem"
        >
          2
        </li>
      </ul>
    </li>
  </ul>
</div>
`;
